### ISSUE TITLE
[TagPicker] onRenderItem expects property name which is documented nowhere

### ISSUE THREAD

**Issue by Fensterbank** [OfficeDev/office-ui-fabric-react#2526]
According to the [Tag Picker examples](http://dev.office.com/fabric#/components/pickers), we could use `getTextFromItem` to define the text to display for a selected item.
This already works this way in `PeoplePicker`, which I used several times.

In `TagPicker` this does not work and the callback result is ignored.
The component always takes the item property `name`:
https://github.com/OfficeDev/office-ui-fabric-react/blob/c4bb4fb265f4c40620110700b78186311e8b8063/packages/office-ui-fabric-react/src/components/pickers/TagPicker/TagPicker.tsx#L25

The Tag Picker example only works, because the example data items have a property called `name`, but the component prop in the following line is not used by the component itself:
https://github.com/OfficeDev/office-ui-fabric-react/blob/07c4a433d922de61db0c174908ec69e6337958ec/packages/office-ui-fabric-react/src/components/pickers/examples/TagPicker.Basic.Example.tsx#L38

**tl;dr:** TagPicker is broken if you don't call `onRenderItem` and if your provided item doesn't have a property `name`.

**Comment by dzearing** [comment]
@joschect

**Comment by joschect** [comment]
Yeah, potentially. Should be an easy fix.

**Comment by Fensterbank** [comment]
I prepared a codepen for this issue and it seem I missunderstood the way `getTextFromItem` works.
https://codepen.io/anon/pen/vJpJar

`getTextFromItems` is working (prefills the input text in the field), but then we still have the problem that `TagPicker` does only show something in the selected item, if the item has the property `name` or if you handle `renderItem` yourself.

![grafik](https://user-images.githubusercontent.com/1684826/29402689-32ccc03e-8336-11e7-94f2-11a26b3ecb20.png)

I think it's an architectural propblem. If this is working as expected, than it should made clear in the documentation.
Because render item expects the existing of `item.name`, it would be possible, but logically nonsense to refer to another property in `getTextFromItem`.

**Comment by joschect** [comment]
This has been fixed, getTextFromItem now correctly reflects that the item returned is of type T, which will match the tag pickers type.

**Comment by Fensterbank** [comment]
@joschect Sorry, but this is wrong.
The bug is still existing and has nothing to do with a type.

A selected item is not rendered correct, if it does not have a `name` prop.
While `getTextFromItem` can be used to select the correct property while typing, we only have two possibilities to make sure the selected item is rendered properly:

- Make sure the item contains a `name` property
- Implement `onRenderItem`, which means we have to take care on rendering the whole node including the remove button.

See this example: https://codesandbox.io/s/x21mw4z7lp

The TagPicker ist still broken if item does not contain a `name` property and getTextFromItem is inconsequent and should also be used when rendering an item.

**Comment by joschect** [comment]
@Fensterbank I am confused, it is working exactly as expected. getTextFromItem is consistent as is the rendering.

In the BasePicker it is defined as `getTextFromItem?: (item: T, currentValue?:string) => string;` where T is passed in through the IBasePickerProps. For the TagPicker it's props extends `IBasePickerProps<ITag>` where ITag is defined as `interface ITag { key: string; name: string;}` that means that the getTextFromItem will have a type of `(item: ITag, currentValue?: string) => string;` 

Similarly the TagPicker expects the items it contains to be of type T, where it is defined as ITag again, so there is no expectation that the item should have a title. Is there documentation somewhere that suggests it should take in a title? 

Additionally I don't feel that getTextFromItem should be used for rendering, it is only to get text based on the searched text to help autocomplete know which text should be autocompleted.

**Comment by Fensterbank** [comment]
@joschect Thanks for clarification

> For the TagPicker it's props extends IBasePickerProps<ITag> where ITag is defined as interface ITag { key: string; name: string;}

Ok, that's true. Then it's clearly stated that the component expects the property `name`.
My prop `title` was just an example. In my use case I have items with localized strings as displayed value.

But then in summary things seem to be inconsequent to me.
What is the purpose of `getTextFromItem` when the only usable human readable text is in the `name` prop?
`getTextFromItem` implies that any attribute can be used as displayed text, while it only works for autocomplete.

And from a user perspective it makes no sense that the text in the autocompletes differs from the text which is displayed in the picker.
`getTextFromItem` is needed for a complete customization, but I don't understand why it is only used for the autocomplete text and not for the text of the not-overwritten `TagItem`.
I cannot imagine someone wants to show the `key` for the autocomplete and the `name` when the item has been selected.

PS: I fixed my issue by importing `TagItem` and overwrite `onRenderItem`, which returns a `TagItem` with another child.
